### PR TITLE
feat: add `sqlcommenter-query-insights` plugin

### PR DESCRIPTION
**Pull request by aqrln**
Implement `@prisma/sqlcommenter-query-insights` plugin. The documentation is currently very generic and doesn't talk about Prisma Postgres because the corresponding functionality isn't yet available in Console. Once it is, the README will be rewritten to be Prisma Postgres centric (although third-party observability providers like Datadog, Sentry etc can also make use of this to provide new cool features for Prisma users).

In the future parameterization will happen in core using the query schema aka DMMF, which is also necessary for query plan caching. For now the plugin temporarily implements its own best-effort parameterization logic, which preserves the hardcoded list of values which are explicitly allowed in certain contexts, and errs on the side of parameterizing everything else.

Closes: https://linear.app/prisma-company/issue/TML-1645/build-ppg-query-insights-plugin

## Summary by CodeRabbit

* **New Features**
  * Added a sqlcommenter plugin that emits parameterized, shape-preserving query insights in SQL comments (supports raw and batched formats).

* **Documentation**
  * Added plugin README and an embedder guide documenting formats, decoding, security considerations, and integration notes.

* **Tests**
  * Added extensive unit, property-based, and end-to-end test suites covering formatting, parameterization, shaping, security, edge cases, and batching.

* **Chores**
  * CI workflow updated to run the new tests and new package metadata added.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: packages/sqlcommenter-query-insights/src/format/format.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+  // Compacted batch
+  const { modelName, action, queries } = queryInfo
+
+  // Parameterize and shape each query in the batch
+  const shapedQueries = queries.map((q) => shapeQuery(parameterizeQuery(q)))
+  const encoded = toBase64Url(JSON.stringify(shapedQueries))
+
+  return `${modelName}.${action}:${encoded}`
+}
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/sqlcommenter-query-insights/src/format/format.ts:50]: compacted batches always have a model name defined

File: packages/client/tests/e2e/sqlcommenter-query-insights/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+// Helper to extract prismaQuery from SQL comment
+function extractPrismaQuery(sql: string): string | null {
+  // Match /*...prismaQuery='...'...*/
+  const match = sql.match(/prismaQuery='([^']*(?:\\'[^']*)*)'/)
+  if (!match) return null
+  // Unescape single quotes and URL-decode
+  return decodeURIComponent(match[1].replace(/\\'/g, "'"))
+}
+
+const PARAM_PLACEHOLDER = { $type: 'Param' }
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client/tests/e2e/sqlcommenter-query-insights/src/index.ts:35]: it's private to the package and should not be exposed in the public api

File: packages/sqlcommenter/README.md [hunk 3]
```diff
@@ -55,6 +55,7 @@ const prisma = new PrismaClient({
 
 - [`@prisma/sqlcommenter-query-tags`](https://www.npmjs.com/package/@prisma/sqlcommenter-query-tags): appends arbitrary tags to all queries within an async context.
 - [`@prisma/sqlcommenter-trace-context`](https://www.npmjs.com/package/@prisma/sqlcommenter-trace-context): appends `traceparent` comments to SQL queries for distributed tracing.
+- [`@prisma/sqlcommenter-query-insights`](https://www.npmjs.com/package/@prisma/sqlcommenter-query-insights): enables query insights for [Prisma Postgres](https://www.prisma.io/postgres).
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/sqlcommenter/README.md:58]: That's very true. I thought it was harmless to already describe it like this here (especially since it makes it concise and clear) and avoids the risk of us forgetting to update this line when we rewrite the README of `sqlcommenter-query-insights` package itself, since this line is in another package's README and is relatively hidden.

File: packages/client/tests/e2e/sqlcommenter-query-insights/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+})
+
+// Test 9: findMany with select - verify select is in payload
+verifyQuery(9, 'findMany with select', {
+  prefix: 'User.findMany',
+  containsInPayload: (p) => {
+    const payload = p as { select?: { id?: boolean; email?: boolean } }
+    return payload.select?.id === true && payload.select?.email === true
+  },
+})
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client/tests/e2e/sqlcommenter-query-insights/src/index.ts]: done

File: packages/sqlcommenter-query-insights/src/parameterize/tests/logical-operators.test.ts [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+                  OR: [{ status: PARAM_PLACEHOLDER }, { status: PARAM_PLACEHOLDER }],
+                },
+              },
+            ],
+          },
+        },
+        selection: { $scalars: true },
+      })
+    })
+  })
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/sqlcommenter-query-insights/src/parameterize/tests/logical-operators.test.ts:372]: Placeholder is an internal detail of the package, not the `parameterize.ts` module. It's fine to import it within the package (especially in the tests for this specific module within the `parameterize` directory), it's not fine to export it outside of the package.

File: packages/client/tests/e2e/sqlcommenter-query-insights/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+}
+
+// Helper to extract prismaQuery from SQL comment
+function extractPrismaQuery(sql: string): string | null {
+  // Match /*...prismaQuery='...'...*/
+  const match = sql.match(/prismaQuery='([^']*(?:\\'[^']*)*)'/)
+  if (!match) return null
+  // Unescape single quotes and URL-decode
+  return decodeURIComponent(match[1].replace(/\\'/g, "'"))
+}
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client/tests/e2e/sqlcommenter-query-insights/src/index.ts:33]: I think either of these is an overkill for the purposes in this test. We don't need to handle nested quotes or escapes here at all.

[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client/tests/e2e/sqlcommenter-query-insights/src/index.ts:33]: I'll accept your suggestion though

File: packages/client/tests/e2e/sqlcommenter-query-insights/_steps.cts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+  test: async () => {
+    await $`pnpm exec prisma -v`
+    await $`tsx src/index.ts`
+    await $`pnpm exec tsc --noEmit`
+  },
+  finish: async () => {
+    await $`echo "done"`
+  },
+  // keep: true, // keep docker open to debug it
+})
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client/tests/e2e/sqlcommenter-query-insights/_steps.cts:20]: this e2e test is executed within a docker container and must contain all steps

File: packages/client/tests/e2e/sqlcommenter-query-insights/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+      const payload = p as { data?: { title?: unknown; content?: unknown } }
+      const data = payload.data
+      return (
+        data !== undefined &&
+        JSON.stringify(data.title) === JSON.stringify(PARAM_PLACEHOLDER) &&
+        JSON.stringify(data.content) === JSON.stringify(PARAM_PLACEHOLDER)
+      )
+    },
+    notContains: ['Secret Title', 'Secret Content'],
+  })
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/client/tests/e2e/sqlcommenter-query-insights/src/index.ts:236]: we have enough unit tests in sqlcommenter-query-insights package itself

File: packages/sqlcommenter-query-insights/docs/embedder-guide.md [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+  const jsonString = fromBase64Url(base64UrlPayload)
+  const payload = JSON.parse(jsonString)
+
+  return {
+    modelName,
+    action,
+    payload,
+    isRaw: false,
+  }
+}
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/sqlcommenter-query-insights/docs/embedder-guide.md:124]: let's not overcomplicate this, it's an illustrative example and not production code

File: packages/sqlcommenter-query-insights/src/parameterize/tests/properties.test.ts [hunk 10]
```diff
@@ -0,0 +1,567 @@
...
+          const result = parameterizeQuery(query) as {
+            selection: { $scalars: boolean; $composites: boolean }
+          }
+          expect(result.selection.$scalars).toBe(scalars)
+          expect(result.selection.$composites).toBe(composites)
+        }),
+        { numRuns: 50 },
+      )
+    })
+  })
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/sqlcommenter-query-insights/src/parameterize/tests/properties.test.ts:566]: that's an overkill to test for that here, but it's a very good point that sqlcommenter infrastructure should make sure that one plugin can't mutate data received by another one; it's outside of scope of this PR though

[the remark this answers is not in the document]
Comment (aqrln) [reply packages/sqlcommenter-query-insights/src/parameterize/tests/properties.test.ts:566]: and if we do that, it doesn't matter if this specific plugin mutates the query

File: packages/client/tests/e2e/sqlcommenter-query-insights/package.json [hunk 11]
```diff
@@ -0,0 +1,24 @@
+{
+  "private": true,
+  "version": "0.0.0",
+  "main": "index.js",
+  "type": "module",
+  "scripts": {},
+  "dependencies": {
+    "@prisma/adapter-better-sqlite3": "/tmp/prisma-adapter-better-sqlite3-0.0.0.tgz",
+    "@prisma/client": "/tmp/prisma-client-0.0.0.tgz",
+    "@prisma/client-runtime-utils": "/tmp/prisma-client-runtime-utils-0.0.0.tgz",
```
Comment (FGoessler) [inline packages/client/tests/e2e/sqlcommenter-query-insights/package.json]: Nit: afaik we don't need the `@prisma/client-runtime-utils` here explicitly.

File: packages/sqlcommenter-query-insights/docs/embedder-guide.md [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+[ModelName.]Action[:Base64UrlEncodedPayload]
+```
+
+### Components
+
+| Component                 | Required        | Description                                                                              |
+| ------------------------- | --------------- | ---------------------------------------------------------------------------------------- |
+| `ModelName`               | No<sup>\*</sup> | The Prisma model name (e.g., `User`, `Post`). Absent for raw queries.                    |
+| `Action`                  | Yes             | The Prisma operation type (see [Actions](#actions) below).                               |
+| `Base64UrlEncodedPayload` | No<sup>\*</sup> | Base64url-encoded JSON containing the parameterized query shape. Absent for raw queries. |
```
Comment (FGoessler) [inline packages/sqlcommenter-query-insights/docs/embedder-guide.md]: Nit: "Absent for raw queries." is redundant info here and in the Model Name.

File: packages/sqlcommenter-query-insights/src/parameterize/parameterize.ts [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+/**
+ * Recursively parameterize a value based on its context.
+ */
+function parameterize(value: unknown, context: Context, key?: string): unknown {
+  // null and undefined always pass through
+  if (value === null || value === undefined) {
+    return value
+  }
+
+  // Handle tagged values
```
Comment (FGoessler) [inline packages/sqlcommenter-query-insights/src/parameterize/parameterize.ts]: Nit: I personally feel like that a lot of comments here are very redundant to the actual code. 🤔

**Review by FGoessler — DISMISSED** [review]
Crazy level of testing ! I like 🫶

File: packages/client/tests/e2e/sqlcommenter-query-insights/package.json (hunk 11 above)
Comment (aqrln) [reply packages/client/tests/e2e/sqlcommenter-query-insights/package.json]: done in d2dac4eb7f8f501bcde9edd44480858e03f0e022

File: packages/sqlcommenter-query-insights/docs/embedder-guide.md (hunk 12 above)
Comment (aqrln) [reply packages/sqlcommenter-query-insights/docs/embedder-guide.md]: addressed in 4cbf516512f30ca5355256a4fa1be4bfaf34c74e

File: packages/sqlcommenter-query-insights/src/parameterize/parameterize.ts (hunk 13 above)
Comment (aqrln) [reply packages/sqlcommenter-query-insights/src/parameterize/parameterize.ts]: addressed in d433b219c84a31881fb0ba2e4bbe1a773e98a653

File: packages/sqlcommenter-query-insights/src/parameterize/parameterize.ts [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+  'disconnect',
+  'set',
+  'create',
+  'createMany',
+  'update',
+  'updateMany',
+  'upsert',
+  'delete',
+  'deleteMany',
+])
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/sqlcommenter-query-insights/src/parameterize/parameterize.ts:32]: addressed in c3b2c6cf7922fdb5b5ca0ad44ceb0eef2d2f9d97

**Review by FGoessler — APPROVED** [review]
